core/server/worker/src/main/java/alluxio/worker/netty/DataServerReadHandler.java

```diff
@@ -391,24 +422,29 @@ public void run() {
       }
     }
 
-    /**
-     * The actual implementation of the runnable.
-     */
     private void runInternal() {
+      boolean eof;  // End of file. Everything requested has been read.
+      boolean cancel;
+      Error error;  // error occured, abort requested.
       while (true) {
         final long start;
         final int packetSize;
         mLock.lock();
         try {
           start = mPosToQueue;
-          long remaining = remainingToQueue();
-          if (tooManyPendingPackets() || remaining <= 0) {
+          eof = mEof;
```

Comment:
This looks like it is locally caching the shared varable (protected by the lock). Is this safe, to then unlock and do some operations on the cached values?